File: rust/src/genesis.rs
use crate::biosquare::{Generation, WorldCreator};
use crate::matrix::Matrix;
use rand::Rng;
use rand_chacha::{rand_core::SeedableRng, ChaCha8Rng};

pub struct DicingGod {
    pub seed: Option<u64>,
    pub initial_density: f64,
}

impl DicingGod {
    pub fn new(seed: Option<u64>, initial_density: f64) -> Self {
        Self {
            seed,
            initial_density,
        }
    }
}

impl WorldCreator for DicingGod {
    fn create(&self, nrows: usize, ncols: usize) -> Generation {
        let mut rng: ChaCha8Rng;

        if let Some(seed) = self.seed {
            rng = ChaCha8Rng::seed_from_u64(seed);
        } else {
            rng = ChaCha8Rng::from_entropy();
        }

        let data: Vec<bool> = (0..nrows * ncols)
            .map(|_| rng.gen_bool(self.initial_density))
            .collect();

        Matrix::from_slice(&data, (nrows, ncols))
    }
}

File: python/src/lifegame/genesis.py
import numpy as np

from .biosquare import Generation


class DicingGod:
    def __init__(self, seed: int | None, initial_density: float) -> None:
        if initial_density < 0:
            raise ValueError("initial density cannot be negative")
        if initial_density > 1:
            raise ValueError("initial density cannot be greater than 1")
        self.seed = seed
        self.initial_density = initial_density

    def create(self, nrows: int, ncols: int) -> Generation:
        rng = np.random.Generator(np.random.PCG64(self.seed))
        init_gen = rng.random(size=(nrows, ncols)) >= self.initial_density
        return init_gen
